packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,319 @@
+    } = e;
+    // prevent default for disabled or loading, but do not impede tab key
+    if (isDisabledOrLoading(this.disabled, this.loading) && key !== 'Tab') {
+      e.preventDefault();
+    } // if input is valid overwrite old value
+    else if (isInputSingleDigit(key)) {
+      e.preventDefault();
+      target.value = key;
+      this.value = getInputValue(this.pinCodeElements);
+      this.emitUpdateEvent();
+
+      if (nextElementSibling) {
+        nextElementSibling.focus();
+      }
+    } // handle alphanumeric keys, allow copy/paste shortcut
+    else if (key?.length === 1 && !(e.ctrlKey || e.metaKey)) {
+      e.preventDefault();
+    } // handle backspace and delete
+    else if (key === 'Backspace' || key === 'Delete') {
+      // transfer focus backward/forward, if the input value is empty
+      if (!target.value) {
+        e.preventDefault();
+        if (key === 'Backspace' && previousElementSibling) {
+          previousElementSibling.value = '';
+          previousElementSibling.focus();
+        } else if (key === 'Delete' && nextElementSibling) {
+          nextElementSibling.value = '';
+          nextElementSibling.focus();
+        }
+      }
+      target.value = '';
+      this.pinCodeElements.map((pinCodeElement) => {
+        if (pinCodeElement === e.target) {
+          pinCodeElement.value = '';
+        }
+        return pinCodeElement;
+      });
+
+      this.value = getInputValue(this.pinCodeElements);
+      this.emitUpdateEvent();
+    } // support native submit behavior
+    else if (key === 'Enter') {
+      if (
+        isWithinForm &&
+        (this.form.querySelectorAll('input').length === 1 || this.form.querySelector('button[type=submit]'))
+      ) {
+        this.form.requestSubmit();
+      }
+    } // workaround since 'Dead' key e.g. ^¨ can not be prevented with e.preventDefault()
+    // workaround for ^ in firefox key: 'Process'
+    else if (key === 'Dead' || key === 'Process') {
+      target.blur();
+      requestAnimationFrame(() => target.focus());
+    }
+  };
+
+  private onPaste = (e: ClipboardEvent): void => {
+    const sanitisedPastedValue = getSanitisedValue(e.clipboardData.getData('Text'), this.length);
+    if (hasInputOnlyDigitsOrWhitespaces(sanitisedPastedValue) && sanitisedPastedValue !== this.value) {
+      this.value = sanitisedPastedValue;
+      this.emitUpdateEvent();
+      this.focusFirstEmptyOrLastElement(sanitisedPastedValue);
+    }
+    e.preventDefault();
+  };
+
+  private emitUpdateEvent = (): void => {
+    this.update.emit({ value: this.value, isComplete: getSanitisedValue(this.value).length === this.length });
+  };
+
+  private focusFirstEmptyOrLastElement = (sanitisedValue: string): void => {
+    if (sanitisedValue.length === this.length) {
+      this.pinCodeElements[sanitisedValue.length - 1]?.focus();
+    } else {
+      this.pinCodeElements[sanitisedValue.length]?.focus();
+    }
+  };
+
+  private validateInitialValue = (): void => {
```

Comment:
Shouldn't this entire function become a utility that can easily be tested?
Because now this function is doing some kind of validation, resetting/santization and event emission.
Also this function is called during render and emits the event? I think the event should be emitted on user interaction and maybe initially if validation fails, otherwise programmatic changes to any property would cause a rerender and potentially emit the event?